JaffParker/vtpayments-test-assignment
# Cj/assignment

Hi Jaff!

  Here is my work for the assignment, I didn’t know GraphQL and React Context/Hooks to manage the state so it took me more time than expected…  I can say I really enjoyed learning GraphQL, pretty straight forward compare to React Context.

  Here is what I decided to implement: 
- A User can own multiple Resells.
- The owner of a Resell is the user who created it.
- It’s not possible to create two resells with the same name.
-  Only the owner of the resell can manage it, so only the owner can add Merchants to his Resell.
- A User can own multiple Merchants.
- A User can own one or multiple Merchant(s) of one or multiple of his Resell(s). 
- When a Resell owner want to add a Merchant, he has to go to the detail page of the Resell and create a Merchant with a name and the email of the user that will owns the Merchant.  If the email provided in the form does not belongs to a user in the database, the form will display an error (same if the merchant name is not available).

##   Here are the steps to run the project:

### For the API
- Start Mysql server: `docker-compose up --build`
- Start the Nest.JS Server to build the DB Schema
- Change the .env `DB_SYNCHRONIZE=false`
- Restart the Nest.JS Server
- Add two users to the Database: 
`docker-compose exec mysql mysql -u user -p vt-terminal`
 
```
INSERT INTO `users` (`email`, `emailConfirmed`, `password`, `profile`, `active`) VALUES ('test@host.com',1,'$2a$10$rYRtv6Ttaa5d3TdYUrpcLOXNMK7mfdrwtimjBpRrVCK/MpIAsWbq6','{\"firstName\":\"Andrey\",\"lastName\":\"Kamozin\"}',1);
```
 
```
INSERT INTO `users` (`email`, `emailConfirmed`, `password`, `profile`, `active`) VALUES ('test2@host.com',1,'$2a$10$rYRtv6Ttaa5d3TdYUrpcLOXNMK7mfdrwtimjBpRrVCK/MpIAsWbq6','{\"firstName\":\"Andrey2\",\"lastName\":\"Kamozin2\"}',1);
```


### For the Frontend
- `yarn start `

## Here are the steps to test:

- Login with: **test@host.com** & **123456**
- Go to Resellers tab  Create a Reseller Click on “More Details” button of the Reseller card
- Add a merchant which you will be the owner (email address: test@host.com)
-  Add a merchant which will belong to the second user (email address: test2@host.com)  (You can’t create two Merchant with the same name, You have to provide an email linked to a user account otherwise you will get an error message in the form)
- Go to Merchants tab to see the merchants you’re owning. There should be only one displayed, the one you created with your email address.
-   Logout and login with **test2@host.com** & **123456**
-  Go to the Merchants tab, you should see the merchant that belongs to test2@host.com you created when you were logged in with test@host.com
- Go to the resellers tab, there should not be any reseller because none has been created for this user yet.  

I hope you will enjoy my work even if the PR is going to be long to review… :D  
Reach out to me if you run into issues to start the project or anything else.

Time spent:
- Backend: 4h
- Frontend: 5h

Have a nice day
César Jeanroy
